File: solutions/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/AC_dfs_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n.cpp
*  Create Date: 2015-03-09 09:39:56
*  Descripton:  Just simulation.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    TreeNode *dfs(vector<int> &preorder, vector<int> &inorder, int ps, int is, int len) {
        if (len <= 0)
            return NULL;
        TreeNode *root = new TreeNode(preorder[ps]);
        for (int i = 0; i < len; ++i) {
            if (inorder[is + i] == root->val) {
                root->left = dfs(preorder, inorder, ps + 1, is, i);
                root->right = dfs(preorder, inorder, ps + i + 1, is + i + 1, len - i - 1);
                break;
            }
        }
        return root;
    }
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        return dfs(preorder, inorder, 0, 0, preorder.size());
    }
};

int main() {
    vector<int> p = {1, 2, 3};
    vector<int> i = {1, 2, 3};
    Solution s;
    s.buildTree(p, i);
	return 0;
}

File: solutions/107.Binary_Tree_Level_Order_Traversal_II/AC_dfs_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_n.py
# Create Date: 2015-03-09 09:34:44
# Usage:       AC_dfs_n.py 
# Descripton:  


# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrderBottom(self, root):
        ret = []
        def dfs(root, level):
            if root:
                if level >= len(ret):
                    ret.append([])
                ret[level].append(root.val)
                dfs(root.left, level + 1)
                dfs(root.right, level + 1)

        dfs(root, 0)
        ret.reverse()
        return ret

